swarm/network/fetcher.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+var searchTimeout = 1 * time.Second
+
+type RequestFunc func(context.Context, *Request) (*discover.NodeID, chan struct{}, error)
+
+// Fetcher is created when a chunk is not found locally. It starts a request handler loop once and
+// keeps it alive until all active requests are completed. This can happen:
+//     1. either because the chunk is delivered
+//     2. or becuse the requestor cancelled/timed out
+// Fetcher self destroys itself after it is completed.
+// TODO: cancel all forward requests after termination
+type Fetcher struct {
+	protoRequestFunc RequestFunc           // request function fetcher calls to issue retrieve request for a chunk
+	addr             storage.Address       // the address of the chunk to be fetched
+	offerC           chan *discover.NodeID // channel of sources (peer node id strings)
+	requestC         chan struct{}
+	skipCheck        bool
+}
+
+type Request struct {
+	Addr        storage.Address  // chunk address
+	Source      *discover.NodeID // nodeID of peer to request from (can be nil)
+	SkipCheck   bool             // whether to offer the chunk first or deliver directly
+	PeersToSkip *sync.Map        // peers not to request chunk from (only makes sense if source is nil)
+}
+
+// FetcherFactory is initialised with a request function and can create fetchers
+type FetcherFactory struct {
+	request   RequestFunc
+	skipCheck bool
+}
+
+// NewFetcherFactory takes a request function and skip check parameter and creates a FetcherFactory
+func NewFetcherFactory(request RequestFunc, skipCheck bool) *FetcherFactory {
+	return &FetcherFactory{
+		request:   request,
+		skipCheck: skipCheck,
+	}
+}
+
+// New contructs a new Fetcher, for the given source. All peers in peersToSkip are not requested to
+// deliver the given chunk. peersToSkip should always contain the peers which are actively requesting
+// this chunk, to make sure we don't request back the chunks from them.
+// The created Fetcher is started the returned function is its fetch function. This function is used by
+// NetStore to fetch a chunk which is not available locally.
+func (f *FetcherFactory) New(ctx context.Context, source storage.Address, peersToSkip *sync.Map) storage.NetFetcher {
+	fetcher := NewFetcher(source, f.request, f.skipCheck)
+	go fetcher.run(ctx, peersToSkip)
+	return fetcher
+}
+
+// NewFetcher creates a new Fetcher for the given chunk address using the given request function.
+func NewFetcher(addr storage.Address, rf RequestFunc, skipCheck bool) *Fetcher {
+	return &Fetcher{
+		addr:             addr,
+		protoRequestFunc: rf,
+		offerC:           make(chan *discover.NodeID),
+		requestC:         make(chan struct{}),
+		skipCheck:        skipCheck,
+	}
+}
+
+func (f *Fetcher) Offer(ctx context.Context, source *discover.NodeID) {
+	select {
+	case <-ctx.Done():
+		return
+	default:
+	}
+
+	select {
+	case f.offerC <- source:
+	case <-ctx.Done():
+	}
```

Comment:
We need this, because otherwise it is possible that the select pushes on offerC although the context is done (because select can choose randomly from the available cases).